Orange/data/table.py

```diff
@@ -539,225 +590,65 @@ def from_url(cls, url):
+        if not (isinstance(key, Variable) or isinstance(key, str)) or key in self.columns:
+            super(Table, self).__setitem__(key, value)
+
+        # assertion: the column is new from here on
+        # the if ordering is important here, since Variable extends str
+        force_meta = False
+        if isinstance(key, Variable):
+            # the variable already exists, don't argue
+            var = key
+        else:  # plain string
+            # we need to construct a new variable (in order of precedence)
+            #  - all numerics: continuous
+            #  - all strings and all distinct: string, meta
+            #  - any string: discrete
+            #  - otherwise: error, can't determine type
+
+            proc_val = value  # for easier type checking even when broadcasting
+            if not isinstance(proc_val, Sequence) or isinstance(proc_val, str):
+                proc_val = [value]
+
+            if all(isinstance(v, Number) for v in proc_val):
+                var = ContinuousVariable(key)
+            elif any(isinstance(v, str) for v in proc_val) and len(set(proc_val)) == len(proc_val):
+                var = DiscreteVariable(key, values=set(proc_val))
+            elif any(isinstance(v, str) for v in proc_val):
+                var = StringVariable(key)
             else:
-                col_idx, values = [col_idx], [value]
-            for value, col_idx in zip(values, col_idx):
-                if not isinstance(value, Integral):
-                    value = self.domain[col_idx].to_val(value)
-                if not isinstance(col_idx, Integral):
-                    col_idx = self.domain.index(col_idx)
-                if col_idx >= 0:
-                    if col_idx < self.X.shape[1]:
-                        self.X[row_idx, col_idx] = value
-                    else:
-                        self._Y[row_idx, col_idx - self.X.shape[1]] = value
-                else:
-                    self.metas[row_idx, -1 - col_idx] = value
-
-        # multiple rows, multiple columns
-        attributes, col_indices = self.domain._compute_col_indices(col_idx)
-        if col_indices is ...:
-            col_indices = range(len(self.domain))
-        n_attrs = self.X.shape[1]
-        if isinstance(value, str):
-            if not attributes:
-                attributes = self.domain.attributes
-            for var, col in zip(attributes, col_indices):
-                if 0 <= col < n_attrs:
-                    self.X[row_idx, col] = var.to_val(value)
-                elif col >= n_attrs:
-                    self._Y[row_idx, col - n_attrs] = var.to_val(value)
-                else:
-                    self.metas[row_idx, -1 - col] = var.to_val(value)
+                raise ValueError("Cannot automatically determine variable type. ")
+
+        self[var] = value
+        # default behaviour is to include this in X, except when meta is set
+        # we can't reasonably separate X and Y here
+        if force_meta:
```

Comment:
A few lines before, where I forgot to set it. Thanks. I've also swapped the string and discrete variable instantiation, it was wrong before.